src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -51,78 +50,90 @@ export class BrowserAdapterEventManager {
     };
 
     private processDeferredEvents(): void {
-        this.deferredEvents = this.deferredEvents.filter(
-            event => this.processEvent(event.eventType, event.eventArgs, true) === false,
-        );
+        const stillDeferred: DeferredEventDetails[] = [];
+        this.deferredEvents.forEach(deferredEvent => {
+            if (this.eventsToApplicationListenersMapping[deferredEvent.eventType]) {
+                try {
+                    this.processEvent(deferredEvent.eventType, deferredEvent.eventArgs);
+                    deferredEvent.deferredResolution.resolveHook(true);
+                } catch (error) {
+                    console.error(error);
+                    deferredEvent.deferredResolution.rejectHook();
+                }
+            } else {
+                stillDeferred.push(deferredEvent);
+            }
+        });
+        this.deferredEvents = stillDeferred;
     }
 
-    private async forwardEventToApplicationListener(
+    private forwardEventToApplicationListener(
         listener: ApplicationListener,
         eventArgs: any[],
     ): Promise<any> {
-        const result = listener(...eventArgs);
+        let result: any;
+        try {
+            result = listener(...eventArgs);
+        } catch (error) {
+            console.error('Error thrown in application listener: ', error);
+            throw error;
+        }
         if (!!result && typeof result.then === 'function') {
             //wrapping application listener promise responses in a 4-minute promise race
             //prevents the service worker going idle before a response is sent
-            await this.promiseFactory.timeout(result, FOUR_MINUTES);
+            return this.promiseFactory.timeout(result, FOUR_MINUTES);
         } else {
             // it is possible that this is the result of a fire and forget listener
             // wrap in 2-minute timeout to ensure it completes during service worker lifetime
-            return this.timeoutFactory.timeoutType === TimeoutType.Alarm
-                ? this.timeoutFactory.createTimeout(
-                      () => Promise.resolve(result),
-                      TWO_MINUTES,
-                      `wrapper-timeout-${Date.now()}`,
-                  )
-                : this.timeoutFactory.createTimeout(() => Promise.resolve(result), TWO_MINUTES);
+            globalThis.setTimeout(() => {
```

Comment:
This block isn't quite right; it needs to be returning a promise which will resolves after two minutes. This is instead returning a promise which resolves immediately, and setting up a `setTimeout` which will create an unrelated Promise after two minutes and then immediately throw it away.

Reply:
To make it easier to test this without actually needing to wait for timeouts, I'd suggest creating a `promiseFactory.delay` which does something like `return new Promise((resolve) => { this.setTimeout(() => resolve(), delayInMillis); });`